### PR TITLE
[1.10] Replaces the missing model with model containing the exception text.

### PR DESCRIPTION
**Pull request by RainWarrior**
screenshot 1
screenshot 2
As a bonus contains simple baked text renderer.
Should probably hide the functionality behind a config option, putting it here to not forget about it.

### CODE REVIEW

**Comment by LatvianModder** [comment]
Why not just keep first line of the error? The rest of the error 99% of the time is the same anyway

**Comment by HenryLoenwind** [comment]
The first line actually contains the least amount of information, unless you mean "the first line of each exception in the 'caused by' chain". That would make some sense.

File: src/main/java/net/minecraftforge/client/model/FancyMissingModel.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    @Override
+    public Collection<ResourceLocation> getTextures()
+    {
+        return ImmutableList.of(font2);
+    }
+
+    @Override
+    public IBakedModel bake(IModelState state, VertexFormat format, Function<ResourceLocation, TextureAtlasSprite> bakedTextureGetter)
+    {
+        //FIXME debug
```
Comment (HenryLoenwind) [inline src/main/java/net/minecraftforge/client/model/FancyMissingModel.java]: ?

Comment (RainWarrior) [reply src/main/java/net/minecraftforge/client/model/FancyMissingModel.java]: yup

**Comment by HenryLoenwind** [comment]
BTW, have you tested this when the Unicode font is enabled?

And when you add that config option, please make it so it defaults to "on" in a dev env.

**Comment by RainWarrior** [comment]
> BTW, have you tested this when the Unicode font is enabled?

[https://github.com/MinecraftForge/MinecraftForge/pull/3328/files#diff-8b9bea265069f8bad55509ef8cc500faR152](https://github.com/MinecraftForge/MinecraftForge/pull/3328/files#diff-8b9bea265069f8bad55509ef8cc500faR152)
[https://github.com/MinecraftForge/MinecraftForge/pull/3328/files#diff-50f9c0e9192f0884735ff9f93401b72dR53](https://github.com/MinecraftForge/MinecraftForge/pull/3328/files#diff-50f9c0e9192f0884735ff9f93401b72dR53)
Lack of unicode support is why I didn't merge baked font renderer in forge when I first wrote it, almost 2 years ago. But I think it'll be useful in the current form anyway.

> And when you add that config option, please make it so it defaults to "on" in a dev env.

Yup, that was on my mind too.

**Comment by RogueLogix** [comment]
i think an option would be best because some modders, including myself, use blocks that we don't really care what it looks like.

**Comment by williewillus** [comment]
@RogueLogic this is just the missing model

**Comment by LexManos** [comment]
Funny, but really the only thing we need is the model name. The exception can be pulled from the log, and the end users wont understand what the exception means.
But this is a good example of how to do text on static models that everyone bitches about.

**Comment by mezz** [comment]
Any update @RainWarrior ?

**Comment by mezz** [comment]
Can this be split into stuff that would be useful for modders, and a test mod that actually adds the crazy debug text?

**Comment by mezz** [comment]
Any update @RainWarrior ?

**Comment by mezz** [comment]
I am closing this since you have not responded in a long time.
If you decide to update this please open a new PR and link back to this one.

**Comment by RainWarrior** [comment]
Reduced the amount of text + added back the cube, after discussion. Reopening
new screenshot

**Comment by RainWarrior** [comment]
final way it should look by default:

**Comment by mezz** [comment]
This text is very unhelpful, can you improve it?


From this issue https://github.com/mezz/JustEnoughItems/issues/703#issuecomment-274070714

**Comment by diesieben07** [comment]
Why is it forge's responsibility if mods have broken models? If it looks bad, it is the fault of the broken mod...

**Comment by mezz** [comment]
I'm about to go down a very slippery slope out of frustration with that sentiment.
_... 3 logical steps later..._
Why not delete the minecraft directory if mods have broken models? If it looks bad, it is the fault of the broken mod...

**Comment by mezz** [comment]
Having the text on the broken model in the inventory, as seen in the screenshot above, is too small and overlapping so it is not helpful. You can't really read it or make sense of it.
It's not Forge's job to make things intentionally crappier, and I seriously doubt it was Fry's intention here. This can be improved.

**Comment by KingLemming** [comment]
This is awful as it stands, though I understand the intentions behind it.

**Comment by diesieben07** [comment]
Yes, I agree it looks terrible. My message above comes across wrongly. Mostly it was a result of me being frustrated with people shipping utterly broken mods like this.

**Comment by covers1624** [comment]
This needs a config, or to only be active in dev.. Users do not need to see this.

**Comment by mezz** [comment]
In-world seems fine, it's readable.   
In-inventory, it may be better shown on a tooltip.

**Comment by williewillus** [comment]
Personally, I doubt if this is necessary at all.
Issues can be easily debugged using the console in dev, and any decent mod should **not** be shipping any builds with model exceptions (cough, cough)...

**Comment by ghost** [comment]
for me, it is a useless thing. noone doesn't need to see that texts on unified items. because we have mouse, and we have waila/theoneprobe like mods that make us understand what item is. as covers1624 say it really needs at least a **config.**

**Comment by LatvianModder** [comment]
Maybe if its possible, move text from item renderer to tooltip?

**Comment by RainWarrior** [comment]
> This needs a config

Guess what? :P

**Comment by HenryLoenwind** [comment]
Ignoring this for 3 months and then coming in here in masses with nothing but negativity? Really, people?

File: src/main/java/net/minecraftforge/common/ForgeModContainer.java [hunk 2]
```diff
@@ -278,12 +279,19 @@ private static void syncConfig(boolean load)
...
         // Client-Side only properties
         propOrder = new ArrayList<String>();
+
         prop = config.get(Configuration.CATEGORY_CLIENT, "replaceVanillaBucketModel", Boolean.FALSE,
                 "Replace the vanilla bucket models with Forges own dynamic bucket model. Unifies bucket visuals if a mod uses the Forge bucket model.");
         prop.setLanguageKey("forge.configgui.replaceBuckets").setRequiresMcRestart(true);
         replaceVanillaBucketModel = prop.getBoolean(Boolean.FALSE);
         propOrder.add(prop.getName());
 
+        prop = config.get(Configuration.CATEGORY_CLIENT, "zoomInMissingModelTextInGui", Boolean.TRUE,
```
Comment (RainWarrior) [inline src/main/java/net/minecraftforge/common/ForgeModContainer.java:289]: Dang, turns out there's 3 places that store the default value for the config property, and not 2.

**Comment by mezz** [comment]
@HenryLoenwind it's because it's in Forge now and seen in-game with mod packs. 
Also I did not believe it would be merged as-is and on by default.

**Comment by LexManos** [comment]
It wasn't merged as is, fry and I talked it through before he merged it. To make it less annoying. as for the text on the item they were SUPPOSED to be restrained in the 16x16 area for the icon until you held the thing in your hand and then it was supposed to be in your face. Either way this is a config option, bitching doesn't help. And yes if mods release broken models they SHOULD be bloody annoying so hopefully they get fixed. It has been proven time and again no matter the nagging we add in dev, modders don't fix anything, we must annoy the user in the correct way with the correct information so they go complain to the right people. As for the slippery slope argument, that's not even worth the effort to yell at you for.

All in all, it seems the things you're complaining about were just a slight issue with the config setting. No need to freak.

**Pull request merged**